backend/typescript/services/implementations/reviewService.ts

```diff
@@ -189,123 +192,117 @@ class ReviewService implements IReviewService {
     };
   }
 
-  async deleteReview(id: string): Promise<void> {
+  // if deleteReview is called without parameter txn, it runs in a try catch block
+  // otherwise, deleteReview should be called inside a try catch block with t as parameter
+  async deleteReview(id: string, txn?: Transaction): Promise<void> {
+    if (txn) {
+      const deleteResult = await this.deleteReviewHelper(id, txn);
+      if (!deleteResult) {
+        throw new Error(`Review id ${id} not found`);
+      }
+      return;
+    }
+
     try {
-      const deleteResult = await this.db.transaction(async (t) => {
-        const reviewToDelete = await PgReview.findByPk(id, {
-          transaction: t,
-          include: [{ all: true, nested: true }],
-        });
+      const deleteResult = await this.db.transaction(async (t) =>
+        this.deleteReviewHelper(id, t),
+      );
+      if (!deleteResult) {
+        throw new Error(`Review id ${id} not found`);
+      }
+    } catch (error) {
+      Logger.error(
+        `Failed to delete Review. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+  }
 
-        if (!reviewToDelete) {
-          throw new Error(`Review id ${id} not found`);
-        }
+  async deleteReviewHelper(id: string, txn: Transaction): Promise<number> {
+    const reviewToDelete = await PgReview.findByPk(id, {
+      transaction: txn,
+      include: [{ all: true, nested: true }],
+    });
+
+    if (!reviewToDelete) {
+      throw new Error(`Review id ${id} not found`);
+    }
 
-        const allBookIds = reviewToDelete.books.map((book: PgBook) => book.id);
-        const deletePublishersAndAuthors = Promise.all(
-          reviewToDelete.books.map(async (book: PgBook) => {
-            // Delete authors (if necessary)
-            book.authors.forEach((author: PgAuthor) => {
-              PgBookAuthor.findAll({
-                where: { author_id: author.id },
-              }).then(async (ret: PgBookAuthor[]) => {
-                const bookAuthorIds = ret.map(
-                  (bookAuthor: PgBookAuthor) => bookAuthor.book_id,
-                );
-                if (
-                  bookAuthorIds.every((bookId) => allBookIds.includes(bookId))
-                ) {
-                  // Delete author
-                  await PgAuthor.destroy({
-                    where: { id: [author.id] },
-                  });
-                }
+    const allBookIds = reviewToDelete.books.map((book: PgBook) => book.id);
+    const deletePublishersAndAuthors = Promise.all(
```

Comment:
None of these operations seem to be using the provided `txn`. Perhaps this explains some of @chriistine's bugs? :thinking: